samza-api/src/main/java/org/apache/samza/startpoint/StartpointVisitor.java

```diff
@@ -24,40 +24,48 @@
    * @param startpointSpecific The {@link Startpoint} that represents the specific offset.
    */
-  void register(SystemStreamPartition systemStreamPartition, StartpointSpecific startpointSpecific);
+  void visit(SystemStreamPartition systemStreamPartition, StartpointSpecific startpointSpecific);
 
   /**
    * Seek to timestamp offset represented by {@link StartpointTimestamp}
    * @param systemStreamPartition The {@link SystemStreamPartition} to seek the offset to.
    * @param startpointTimestamp The {@link Startpoint} that represents the timestamp offset.
    */
-  void register(SystemStreamPartition systemStreamPartition, StartpointTimestamp startpointTimestamp);
+  default void visit(SystemStreamPartition systemStreamPartition, StartpointTimestamp startpointTimestamp) {
+    throw new UnsupportedOperationException("StartpointTimestamp is not supported.");
+  }
 
   /**
-   * Seek to earliest offset represented by {@link StartpointEarliest}
+   * Seek to earliest offset represented by {@link StartpointOldest}
    * @param systemStreamPartition The {@link SystemStreamPartition} to seek the offset to.
-   * @param startpointEarliest The {@link Startpoint} that represents the earliest offset.
+   * @param startpointOldest The {@link Startpoint} that represents the earliest offset.
    */
-  void register(SystemStreamPartition systemStreamPartition, StartpointEarliest startpointEarliest);
+  default void visit(SystemStreamPartition systemStreamPartition, StartpointOldest startpointOldest) {
+    throw new UnsupportedOperationException("StartpointOldest is not supported.");
+  }
 
   /**
-   * Seek to latest offset represented by {@link StartpointLatest}
+   * Seek to latest offset represented by {@link StartpointUpcoming}
    * @param systemStreamPartition The {@link SystemStreamPartition} to seek the offset to.
-   * @param startpointLatest The {@link Startpoint} that represents the latest offset.
+   * @param startpointUpcoming The {@link Startpoint} that represents the latest offset.
    */
-  void register(SystemStreamPartition systemStreamPartition, StartpointLatest startpointLatest);
+  default void visit(SystemStreamPartition systemStreamPartition, StartpointUpcoming startpointUpcoming) {
+    throw new UnsupportedOperationException("StartpointUpcoming is not supported.");
+  }
 
   /**
-   * Bootstrap signal represented by {@link StartpointBootstrap}
+   * Bootstrap signal represented by {@link StartpointCustom}
    * @param systemStreamPartition The {@link SystemStreamPartition} to seek the offset to.
-   * @param startpointBootstrap The {@link Startpoint} that represents the bootstrap signal.
+   * @param startpointCustom The {@link Startpoint} that represents the bootstrap signal.
    */
-  void register(SystemStreamPartition systemStreamPartition, StartpointBootstrap startpointBootstrap);
+  default void visit(SystemStreamPartition systemStreamPartition, StartpointCustom startpointCustom) {
```

Comment:
How can we ensure that implementations of StartpointCustom are serializable into the store? Also, will visitor implementations need to cast this to the actual concrete custom type being expected?

Reply:
I have a unit test for this in `TestStartpointSerde`, though I forgot to rename `testStartpointBootstrapSerde` to `testStartpointCustomSerde`, so it's good you commented here. Will change the test method name.

Reply:
Your test case has a couple of primitive fields, which JSON can handle easily, but when there is an option to do arbitrary structures, some cases might not work well with JSON. Would it be useful to doc that JSON will be used and that implementations should be simple POJOs or something?

Reply:
Good point. Will add documentation for it. Thanks.